Adapt difficulty to player performance

guyonvarch opened the pull request.

Commit:
Simplify logic on getSpeed functions

Commit:
Give more control over speed

Commit:
Remove unused trick feature

Commit:
Make the skiier invincible for some time after a hit

Commit:
Use or in collision condition to fasten the test

Commit:
WIP prevent object to appear before and after jump

Commit:
Normalize drop rate by speed ratio

Avoid overspawning when stopped or underspawning when jumping.

Commit:
Uniformize player speed input with direction

No matter if it’s coming from keyboard or Senso.

Commit:
Use y position to show objects in front of others

Commit:
Slow down when hitting snow

Commit:
Prevent droping more objects if skier is slow

Commit:
Use nix and esbuild

Commit:
Use const and let instead of var

Commit:
Use ES6 modules and remove sugar dependency

Commit:
Use root import and use esbuild to serve files

Commit:
Move js files to src

Commit:
Remove remaining usage of that

Commit:
Derive isMoving from speed

Commit:
Add basic physic support for moving the player

Commit:
Add plausible physic support

Acceleration is based on:

- skies direction (the more the skies are facing south, the more the
  acceleration is important, force toward skies direction)
- skies going against current speed (force against speed),
- friction to limit acceleration (force against speed),

Commit:
Check TS files with tsc

Commit:
Remove unmaintained tests

It’s causing trouble when I’m searching for code, and it has not been
adapted to work with our last changes (at least mines).

Commit:
Migrate JS to TS

Commit:
Fix limiting side directions

Commit:
Fix distance metrics

Commit:
Extract sorting objects to draw in function

Commit:
Fix clearing hitbox when debugging

Commit:
Prevent spawning when stopped

Use a random from 1 instead of 0 to prevent spawning objects when the
skier is going very slow (0.00001 for ex.), but not at a null speed.

Commit:
Remove z position in sprites

Commit:
Fix generating jumps and snows

Commit:
Simplify computation of hitbox

Commit:
Improve collisions precision

Commit:
Replace guid by counter id

This is simpler and enough for our local needs.

Commit:
Fix collision when using zoom

Commit:
Allow similar objects to be grouped together

Commit:
Assure a clear landing area after a jump

Commit:
Allow an object to hit the skier multiple times

This is convenient with snowboarders for example.

Commit:
Move skier with chunks to prevent speed boost

Commit:
Adjust difficulty to performance

Commit:
Show confidence boost when debug is on

Commit:
Boost confidence with jumps

Commit:
Allow the monster to eat the skier even if invincible

Otherwise, it makes the monster to follow the skier, that looks a bit
strange.

Commit:
Add trees on the side of the screen

Commit:
Put obstacles toward the place the skier is going to

This ensures if no one is playing for example, that the performance is
worse than if someone is playing not so good.

Commit:
Remove unused sprite code

Commit:
Zoom in and out according to confidence

Commit:
Add keyboard inertia

It make it more interesting it to play with the keyboard.

Commit:
Spawn boarders with different speeds and also below viewport

Commit:
Allow to specify multiple hitboxes per sprite

Commit:
Make collisions more precise

Use multiple rectangles when necessary.

Commit:
Optimise collisions

Commit:
Get rid of canvasRenderingContext2DExtensions

Commit:
Adjust game to different screen sizes

Try to make the game look the same no matter the size of the screen.

Commit:
Fix making bundle

Commit:
Increase drop rates

Commit:
Harden canvas initialization

Prevent having null values initially. Don’t adapt sizes when screen size
is changed, because we don’t proporly handle it anyway.

Commit:
Reduce obstacles spawning

Commit:
Use back an image reduction of 33 %

Commit:
Make the game looks the same no matter the size reduction

Commit:
Create jump landing zone only if taken

Without this change, if we would have increase the number of jumps, it
would have increased the number of clear landing areas as well. This
would result in some parts of the area being blank.

Commit:
Free area before jumps

Commit:
Prevent boarders to form groups

Snowboarders could not go too far away from the skier, which could cause
them to group together 600 px after the skier. I noticed that by
settings the zoom level in game.ts to be a fixed 0.5.

Now, just make the snowboarder to go after his own position.

Commit:
Use vec2 for positions

Also simplify checking collisions without callbacks.

knuton commented on makefile:
```diff
@@ -4,30 +4,21 @@ CFLAGS = -c -g -D $(VERSION)
-	cd sprites && mogrify -resize 33% *.png
-	# https://pngquant.org/
+	cd sprites && mogrify -resize 33% *.png # Should be equal to config.spriteSizeReduction
+	https://pngquant.org/
```
I noticed that this URL went from being a comment to being a command :smile: (As the dependency on `pngquant` is made explicit in the flake file now, may as well drop the URL comment.)

I think this may be the reason the sprites have a multiple of the previous filesize even though they are now the same pixel size again. Maybe worth reverting to the original files now that the pixel size remains the same?

guyonvarch replied:
Ah, I was doing to many things to take care of the error when re-generating images, thanks: https://github.com/dividat/skifree.js/pull/9/commits/cac9320e9fe6514367346651ff1dbeab1e7406f8

knuton commented on index.html:
```diff
@@ -16,9 +16,8 @@
         <script src="PlayEGI.js"></script>
         <script src="PlayEGIHelpers.js"></script>
-        <script src="dist/skifree.min.js"></script>
+        <script src="dist/skifree.js"></script>
```
I think this will reference the wrong file in a non-dev build?

guyonvarch replied:
Producing a non min named as min in dev to fix that quickly: https://github.com/dividat/skifree.js/pull/9/commits/ea6e4eaae7859430c369620c947c0f7b9dc6299d

knuton commented:
I think the title "Adapt difficulty to player performance" is a bit of an understatement for this PR now. :joy:

Commit:
Fix image generation command

And revert to existing sprite sizes.

Commit:
Use optimized ski file in index.html

guyonvarch commented:
> I think the title "Adapt difficulty to player performance" is a bit of an understatement for this PR now.

Yeah, it was the driver for this change, but I went further in the end. :sweat_smile:

knuton commented:
> Yeah, it was the driver for this change, but I went further in the end.

It's Ski 3 and it's great!

knuton commented on src/config.ts:
```diff
@@ -0,0 +1,62 @@
+    }
+  },
+  skier: {
+    activeSensoRatio: 1.5 / 3,
+    directionAmplitudeRatio: 0.6,
+    verticalPosRatio: 0.15,
```
It would be nice to explain these, as I think they are quite crucial for the game's behaviour.

I have read the code at use site, but I am not sure I understand what they do.

guyonvarch replied:
https://github.com/dividat/skifree.js/pull/9/commits/e99c800e71e1fb49587e54bd4b38fa870f275023

Commit:
Make the monster more interesting

Make it follow the player according to his speed, and to be tired and
stop the chase, either after some time, or after the skier makes a jump.

Commit:
Reduce monster apparition and difficulty

Playing on the flex, it became too hard.

Commit:
Reduce active senso amplitude

Commit:
Explain skier config

knuton committed:
Fix wheelchair setting

The sensitivity adjustment for the wheelchair flag had not been
reapplied at the correct place after the switch to directions expressed
in radians.

This applies the senstivity adjustment to the horizontal deviation
before computing the angle. Sensitivity is set to one of two values in
the config based on the wheelchair flag, it would be possible to let
users choose from a continuous range instead.

knuton committed:
Revert "Reduce monster apparition and difficulty"

This reverts commit 81a9cbb039ceab153c372a3fb9dffecede2f67a6.

knuton committed:
Revert "Make the monster more interesting"

This reverts commit 4473f8cb4cf6e0659c4e92a93fc8954fed4cf84b.

knuton committed:
Remove debug log

knuton commented:
Test checklist for latest changes

- [x] Retest wheelchair mode
- [x] Test adjustment across screen resolutions
- [x] Test new smaller amplitude range (4ec54c7)
- [x] Test difficulty of new chase mode on Senso

knuton committed:
Revert "Reduce active senso amplitude"

This reverts commit 4ec54c7e70ee70e47a285ed1844d53cacd5b8770.

knuton commented:
The changes in 4473f8c/81a9cbb are really cool, but require more balancing. It seems very hard (very dependent on luck) to actually escape the monster. I think that's probably even fine, but maybe then the likelihood of a monster appearing should be low initially and grow with the skier's confidence value or something similar that limits effects on poor performing players. I reverted this change to unblock the myriad of improvements.

I also reverted the reduction in amplitude (4ec54c7) because it made the controls quite volatile.

knuton committed:
Update makefile 'help'

knuton reviewed (approved):
Fantastic improvement!
